include/deal.II/grid/cell_id_translator.h

```diff
@@ -0,0 +1,277 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2021 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_cell_id_translator_h
+#define dealii_cell_id_translator_h
+
+#include <deal.II/base/exceptions.h>
+#include <deal.II/base/geometry_info.h>
+
+#include <deal.II/grid/cell_id.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace internal
+{
+  /**
+   * A class that helps to translate between CellIDs and globally uniquely
+   * indices. The resulting index space might be non-contiguous in the case
+   * of locally refined meshes.
+   *
+   * The following code snippet how to set up an IndexSet object for active
+   * cells, using this class:
+   * @code
+   * // set up translator
+   * CellIDTranslator<dim> translator(tria.n_global_coarse_cells(),
+   *                                  tria.n_global_levels());
+   *
+   * // initialize the index set with the correct size
+   * IndexSet index_set(translator.size());
+   *
+   * // convert IDs to indices and add them to the index set
+   * for(const auto & cell : tria.active_cell_iterators())
+   *   index_set.add_index(translator.translate(cell));
+   * @endcode
+   *
+   * @note The definition of the indices are not relevant.
+   */
+  template <int dim>
+  class CellIDTranslator
+  {
+  public:
+    /**
+     * Constructor taking the number of global coarse cells and number of
+     * global levels of a triangulation.
+     */
+    CellIDTranslator(const types::global_cell_index n_coarse_cells,
+                     const types::global_cell_index n_global_levels);
+
+    /**
+     * Return maximum number of cells, i.e., in the case the triangulation
+     * would be refined globally.
+     */
+    types::global_cell_index
+    size() const;
+
+    /**
+     * Convert a @p cell of type CellAccessor or DoFCellAccessor to an index.
```

Comment:
Shouldn't these be iterators and not accessors?

Reply:
If they are iterators - I would rather this be
```cpp
template <typename Accessor>
types::global_cell_index
translate(const TriaRawIterator<Accessor> &cell)
```
or `translate(const TriaIterator<Accessor> &cell)` for a little more type safety.